### PR TITLE
Fix linter warnings when shell magic assignment statements are present

### LINKED ISSUE #455: Fix linter warnings raised when ! magic output is assigned to a variable
Given below code snippet

```Python
flake8_version = !pip list 2>&1 | grep flake8

if flake8_version:
    print(flake8_version)
```

Mypy, flake8 and pylint raise `undefined name 'flake8_version'`

### PR DESCRIPTION
**Pull request by girip11**
closes #455

### CODE REVIEW

**Comment by girip11** [comment]
I would like to add more documentation to the `ShellCommandHandler` class.

File: nbqa/handle_magics.py [hunk 1]
```diff
@@ -148,18 +184,9 @@ def is_ipython_magic(source: str) -> bool:
...
-            # contains any `get_ipython` function call.
-            # If the statement is a valid ipython magic, then ipython2python
-            # will transform using another `get_ipython()` function call.
-            src_count = source.count("get_ipython()")
-            warnings.simplefilter("ignore", DeprecationWarning)
-            # see https://github.com/nbQA-dev/nbQA/issues/459
-            return ipython2python(source).count("get_ipython()") == (1 + src_count)
+        return MagicHandler._ipython2python(source).count(
+            "get_ipython()"
+        ) > source.count("get_ipython()")
```
Comment (MarcoGorelli) [inline nbqa/handle_magics.py:189]: barring refactoring, the only change here is
```diff
- == 1 + src_count
+ > src_count
```

(note to self :smile: )

File: nbqa/handle_magics.py [hunk 2]
```diff
@@ -221,6 +248,89 @@ class HelpMagicHandler(MagicHandler):
...
     """Handle ipython magic containing !."""
 
+    _assign_shell_output = False
+    _cmd = ""
+    _var = ""
+
+    _EXTRACT_COMMAND_PATTERN: str = r"getoutput\('(.*)'\)"
+    _EXTRACT_VARNAME_PATTERN: str = r"(\w+)\s*="
+    _REPLACE_MAGIC_TEMPLATE: str = r'{var} = ascii(""" {command} """)  # {token}'
+    _RESTORE_MAGIC_REGEX_TEMPLATE: str = r"{var}\s*=\s*ascii.*{token}"
```
Comment (MarcoGorelli) [inline nbqa/handle_magics.py:258]: wow, nice idea!

**Comment by MarcoGorelli** [comment]
Awesome work! Let me know when it's ready for review - further documenting/commenting can always come in a separate PR if you want

**Comment by girip11** [comment]
>  Let me know when it's ready for review - further documenting/commenting can always come in a separate PR if you want

@MarcoGorelli : I have updated the issue #455 with the approaches I had thought of. The problem with those approaches is that when we have a long shell command that itself takes up 70-100 as its line length, then black will keep reformatting the notebook. That is the reason I didn't make the PR ready for review. Do you think we should go ahead with this PR?

**Comment by MarcoGorelli** [comment]
> The problem with those approaches is that when we have a long shell command that itself takes up 70-100 as its line length, then black will keep reformatting the notebook.

Hey - I couldn't reproduce this, I did
```bash
(nbqa-dev) mgorelli@m-gorelli02:~/nbqa-dev$ python -m nbqa black tests/data/notebook_with_indented_magics.ipynb --nbqa-mutate
reformatted tests/data/notebook_with_indented_magics.ipynb
All done! ✨ 🍰 ✨
1 file reformatted.
```
and the relevant cell in the notebook has become
```python
flake8_version = ascii(
    """ pip list 2>&1                                                  | grep                                                            flake8 """
)  # 0x877481CE
```

If I then run ` python -m nbqa black tests/data/notebook_with_indented_magics.ipynb --nbqa-mutate` again, then it doesn't reformat, even though the string is now really long.

So it looks like something's gone wrong during substituion - however, maybe instead of making the replacement
```
r'{var} = ascii(""" {command} """)  # {token}'
```
we could do something like
```
r'{var} = ascii(\n    """ {command} """\n)  # {token}'
```
? I haven't tried this yet, but I think that `black` is OK with long strings if they're on their own line?

**Comment by girip11** [comment]
> `r'{var} = ascii(\n    """ {command} """\n)  # {token}'`

I have to check if black does not reformat the code when the entire line can fit within the configured line length.

**Comment by girip11** [comment]
I have confirmed that black reformats the code every time when I used `r'{var} = ascii(\n """ {command} """\n) # {token}'` and the actual line length was less than the max line length configured for black.

I need to do some more thinking it seems. Marking this PR as draft.

**Comment by MarcoGorelli** [comment]
> > `r'{var} = ascii(\n    """ {command} """\n)  # {token}'`
> 
> I have to check if black does not reformat the code when the entire line can fit within the configured line length.

Oh yeah, you're right...this is a tricky one to deal with, I think we _may_ have to treat it as a known limitation. The person who reported the bug with shell magic assignment was only using `black` and `pyupgrade` anyway IIRC - in fact, autoformatting seems to be a much more common use-case for nbqa than linting.

One suggestion would be that we _could_ make a 0.5.0 release, with support for shell magic assignment (though with this known limitation for linter warnings)

**Comment by girip11** [comment]
@MarcoGorelli  We could go ahead and release next version of nbqa. I will keep making progress on this PR. As you pointed out, formatting tools are majorly used with nbqa. So we could treat the issue with the linter as a known limitation.

**Comment by girip11** [comment]
closing this pull request and raising a new one when ready

**Pull request closed without merging**